Sming/Libraries/ModbusMaster/samples/generic/app/application.cpp

```diff
@@ -0,0 +1,125 @@
+
+#define RS485_RE_PIN 15
+#define MB_SLAVE_ADDR 1
+#define SLAVE_REG_ADDR 1
+
+Timer mbLoopTimer;
+
+ModbusMaster mbMaster;
+HardwareSerial Serial1(UART1);
+
+uint16_t globalSeconds = 0;
+
+void mbLoop()
+{
+	globalSeconds++;
+
+	uint8_t numberOfRegistersToWrite = 1;
+	uint8_t bufferPosition = 0;
+	mbMaster.begin(MB_SLAVE_ADDR, Serial);
+	mbMaster.setTransmitBuffer(bufferPosition, globalSeconds);
+	mbMaster.writeMultipleRegisters(SLAVE_REG_ADDR, numberOfRegistersToWrite);
+
+	uint8_t nrOfRegistersToRead = 1;
+	uint8_t mbResult = mbMaster.readHoldingRegisters(SLAVE_REG_ADDR, nrOfRegistersToRead); //see also readInputRegisters
+	uint16_t result = 0;
+
+	if(mbResult == mbMaster.ku8MBSuccess) {
+		/*
+		for(i = 0; i < nrOfRegisters; i++) {
+			buffer[i] = mbMaster.getResponseBuffer(i);
+			Serial1.printf("Reg %d: %d\r\n", i, buffer[i]);
+		}
+		*/
+		debugf("Data from slave: %d", mbMaster.getResponseBuffer(0));
+	} else {
+		debugf("Res err: %d", result);
+	}
+
+	mbMaster.clearResponseBuffer();
+}
+
+void init()
+{
+	pinMode(RS485_RE_PIN, OUTPUT);
+	digitalWrite(RS485_RE_PIN, 0);
+	Serial.begin(115200, SERIAL_8N1, SERIAL_FULL);
+	Serial1.begin(SERIAL_BAUD_RATE, SERIAL_8N1,
+				  SERIAL_TX_ONLY); // 115200 by default, GPIO1,GPIO3, see Serial.swap(), HardwareSerial
+	Serial1.systemDebugOutput(true);
+
+	Debug.setDebug(Serial1);
+	Debug.initCommand();
+	Debug.start();
+	Debug.printf("This is the debug output\r\n");
+
+	mbMaster.preTransmission(preTransmission);
+	mbMaster.postTransmission(postTransmission);
+	mbMaster.logReceive(mbLogReceive);
+	mbMaster.logTransmit(mbLogTransmit);
+
+	mbLoopTimer.initializeMs(1000, mbLoop).start();
+}
+
+void preTransmission()
+{
+	digitalWrite(RS485_RE_PIN, 1);
+	delayMilliseconds(2);
+}
+
+void postTransmission()
+{
+	delayMicroseconds(500);
+	digitalWrite(RS485_RE_PIN, 0);
+}
+
+void mbLogReceive(const uint8_t* adu, size_t aduSize, uint8_t status)
+{
+	if(status != mbMaster.ku8MBSuccess) {
+		switch(status) {
+		case mbMaster.ku8MBIllegalFunction:
+			debugf("MB Illegal Function");
+			break;
+		case mbMaster.ku8MBIllegalDataAddress:
+			debugf("MB Illegal Address");
+			break;
+		case mbMaster.ku8MBIllegalDataValue:
+			debugf("MB Illegal Data Value");
+			break;
+		case mbMaster.ku8MBSlaveDeviceFailure:
+			debugf("MB Slave Device Failure");
+			break;
+		case mbMaster.ku8MBInvalidSlaveID:
+			debugf("MB Invalid Slave ID");
+			break;
+		case mbMaster.ku8MBInvalidFunction:
+			debugf("MB Invalid function");
+			break;
+		case mbMaster.ku8MBResponseTimedOut:
+			debugf("MB Response Timeout");
+			break;
+		case mbMaster.ku8MBInvalidCRC:
+			debugf("MB Invalid CRC");
+			break;
+		case mbMaster.ku8MBResponseTooLarge:
+			debugf("MB Response too large");
+			break;
+		}
+		debugf("ADU Size: %d, status: %d ", aduSize, status);
+	}
+	debugf("\r\n");
+}
+
+void mbLogTransmit(const uint8_t* adu, size_t aduSize)
+{
+	uint8_t i = 0;
+	while(i < aduSize) {
+		debugf("ADU[%d]: %d, ", i, adu[i]);
+		i++;
+	}
+	debugf(" ");
```

Comment:
You could replace this with `debugHex(INFO, "ADU", adu, aduSize);`

Reply:
Did grep -r "debugHex" from /opt/Sming with output of 0 results?

Reply:
So sorry, it's `debug_hex`. I was getting confused with `m_printHex` I think, which it uses.

Reply:
Done.